db/db_impl_compaction_flush.cc

```diff
@@ -1367,40 +1415,53 @@ Status DBImpl::BackgroundFlush(bool* made_progress, JobContext* job_context,
     return status;
   }
 
-  ColumnFamilyData* cfd = nullptr;
+  autovector<ColumnFamilyData*> cfds;
+  autovector<MutableCFOptions> mutable_cf_options_list;
+  autovector<uint64_t> memtable_ids;
   while (!flush_queue_.empty()) {
     // This cfd is already referenced
-    auto first_cfd = PopFirstFromFlushQueue();
-
-    if (first_cfd->IsDropped() || !first_cfd->imm()->IsFlushPending()) {
-      // can't flush this CF, try next one
-      if (first_cfd->Unref()) {
-        delete first_cfd;
+    const auto& flush_req = PopFirstFromFlushQueue();
+
+    for (auto& iter : flush_req) {
+      auto cfd = iter.first;
+      if (cfd->IsDropped() || !cfd->imm()->IsFlushPending()) {
+        // can't flush this CF, try next one
+        if (cfd->Unref()) {
+          delete cfd;
+        }
+        continue;
       }
-      continue;
+      cfds.push_back(cfd);
+      mutable_cf_options_list.emplace_back(*cfd->GetLatestMutableCFOptions());
+      memtable_ids.push_back(iter.second);
+      job_context->superversion_contexts.emplace_back(
```

Comment:
I didn't find where we changed it from `superversion_context` to `superversion_contexts` but I don't like the idea that JobContext is not self-sufficient, and it relies on the same array order with cfds outside JobContext to be able to operatable. Can we do, for example a map from cfd to SuperVersionContext instead?

Reply:
The change was introduced by #3949.
I had discussion with @ajkr. I have been under the impression that we should avoid using containers other than autovector inside db mutex to reduce memory allocation pressure. Please correct me if this is wrong.

Reply:
Let me try if we can use a map and reserve space for the map outside a critical section.

Reply:
If memory allocation is the problem, you can make it a autovector of pair<CFD, SuperVersionContext>, or something like that. The thing I don't like is that, JobContext itself has some SuperVersion objects for which to know which CF they are for, we only rely on the order of an array object in some remote place.

Reply:
We might be able to alleviate the memory issue. We peek the top of the queue first in order to determine the size of the vectors. Then we perform memory allocation outside critical section.

Reply:
If we can avoid one allocation, that is even better:)

Reply:
btw, we cannot use autovector<SuperVersionContext> because it requires copy constructor of SuperVersionContext and SuperVersion not to be deleted.

Reply:
Oh it is currently std::vector. Then what's the different about allocation between vector and map?

Reply:
Ok. I start to consider changing it back to a single SuperVersionContext instead of a vector, just the way we did in the past before #3949. Then `BackgroundCallFlush` should allocate all the data structures needed, e.g. `BGFlushArg`s, `JobContext`s, etc. If we do this, we will have multiple `JobContext`s for an atomic flush. This is different from my motivation for #3949, but is easier to implement and not hard to understand.

Reply:
You can keep it a vector of a pair, of a map. It's almost the same as what it is now, right?

Reply:
I probably missed something, but am wondering - if `SuperVersionContext` is associated with a single column family and we need to remember which one, why not simply give it a `ColumnFamilyData*`?

Reply:
When we create a `SupverVersionContext`, we may not know which column family it is associated with. Later we call `ColumnFamilyData::InstallSuperVersion` to associate a `SupverVersion` with a `ColumnFamilyData`.
If we add a `ColumnFamilyData*` to `SuperVersionContext`, it will look strange.
* What value should we give it when we first create a `SuperVersionContext`? `nullptr`? Then I guess we need to add many asserts to make sanitizer happy. We also need to be careful to check its value before trying to dereference.
* `SupverVersionContext`->`SuperVersion`->`Version`->`ColumnFamilyData`. This information is already there even if we do not add a `ColumnFamilyData*` to `SupverVersionContext`.

Reply:
Got it thanks. If column family is available already through that chain, can you access it in that way?

Reply:
@siying @ajkr Thanks for the comments.
I think current `JobContext` is self-contained based on the following.
* Before calling `ColumnFamilyData::InstallSuperVersion`, there is no correlation between `ColumnFamilyData` and `SuperVersionContext`.
* After calling `ColumnFamilyData::InstallSuperVersion`, we can use the following to map `SuperVersionContext` to `ColumnFamilyData`:
`SuperVersionContext` -> `SuperVersion` -> `Version` ->`ColumnFamilyData`.
Back to the suggestion of @siying. If we use a `map` or `vector<pair>`, then the question arises when we create `JobContext` object. What column family does this `SuperVersionContext` belong to? If the answer is not clear, then we need to assign some value, which can complicate the logic.
Based on the above, I suggest that we keep the current `JobContext`.